### PR TITLE
refclock / audio resampling

### PR DESCRIPTION
**Pull request by mk01**

### CODE REVIEW

File: xbmc/cores/VideoPlayer/VideoRenderers/RenderManager.cpp [hunk 1]
```diff
@@ -1400,24 +1399,5 @@ bool CRenderManager::GetStats(int &lateframes, double &pts, int &queued, int &di
...
-  if (diff < 0.01)
-  {
-    m_clockSync.m_enabled = true;
-  }
-  else
-  {
-    m_clockSync.m_enabled = false;
-    m_dvdClock.SetSpeedAdjust(0);
-  }
+  m_clockSync.m_enabled = m_fps != 0.0f;
```
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoRenderers/RenderManager.cpp]: this is wrong. did you umderstand what it does?

File: xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp [hunk 2]
```diff
@@ -2453,7 +2453,7 @@ CSampleBuffer* CActiveAE::SyncStream(CActiveAEStream *stream)
         CLog::Log(LOGDEBUG,"ActiveAE::SyncStream - average error %f, last average error: %f", error, stream->m_lastSyncError);
         stream->m_lastSyncError = error;
       }
-      else
+      else if (stream->m_lastSyncError / error < .0)
```
Comment (FernetMenta) [inline xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp]: with this it takes ages for audio to get started?

File: xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp [hunk 3]
```diff
@@ -2462,6 +2462,13 @@ CSampleBuffer* CActiveAE::SyncStream(CActiveAEStream *stream)
         CLog::Log(LOGDEBUG,"ActiveAE::SyncStream - average error %f below threshold of %f", error, 30.0);
       }
     }
+    else
+    {
+      static double lerr;
```
Comment (FernetMenta) [inline xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp]: :) seriously?

File: xbmc/cores/VideoPlayer/VideoRenderers/RenderManager.cpp (hunk 1 above)
Comment (mk01) [inline xbmc/cores/VideoPlayer/VideoRenderers/RenderManager.cpp]: yep, my bad. changed to what original intention was.

File: xbmc/cores/VideoPlayer/VideoRenderers/RenderManager.cpp [hunk 4]
```diff
@@ -1400,6 +1400,12 @@ bool CRenderManager::GetStats(int &lateframes, double &pts, int &queued, int &di
 
 void CRenderManager::CheckEnableClockSync()
 {
+  if (CSettings::GetInstance().GetBool(CSettings::SETTING_VIDEOPLAYER_USEDISPLAYASCLOCK))
+  {
+    m_clockSync.m_enabled = true;
```
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoRenderers/RenderManager.cpp]: this is wrong. do you know what m_clockSync.m_enabled does?

File: xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp (hunk 2 above)
Comment (mk01) [inline xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp]: one error interval at most. so perhaps yes, ages. on the other hand it allows (together with the convergence) to keep stable resampling even if refClock is wrong.
but beside robustness test it is stupid scenario anyhow... removed.

File: xbmc/cores/VideoPlayer/VideoRenderers/RenderManager.cpp (hunk 4 above)
Comment (mk01) [inline xbmc/cores/VideoPlayer/VideoRenderers/RenderManager.cpp]: @FernetMenta 

provides the gap/error calc for passthrough, but also it shifts the dvd audio clock which provides the callbacks for getting streams clock (if I remember correctly).

basically when playing for instance 23.976 on screen at 60, rr gets quite erratic, without getting stable. allowing that offset calculation rr gets stable quite fast. if you give me any other pointer I will gladly go again with research...

Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoRenderers/RenderManager.cpp]: you are wrong. it calculates the offset between actual pts and vblank. assume clock is started 20ms after vblank has occurred. a/v sync would be off by 20ms. the logic here calculates this offset but this won't work with a moving target, means if refresh rate does not exactly match video fps.

it is not related to passthrough or rr for PCM.

Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoRenderers/RenderManager.cpp]: @mk01 not sure why you think rr would not stabilize. this is playing 23.976 at a 60hz display:

```
12:19:01 T:123145308819456  NOTICE: ----- error: -6.232591, rr: 0.996408, prop: -0.003116, int: 0.000523
12:19:02 T:123145308819456  NOTICE: ----- error: -3.835275, rr: 0.997607, prop: -0.001918, int: 0.000523
12:19:03 T:123145308819456  NOTICE: ----- error: -2.754601, rr: 0.998147, prop: -0.001377, int: 0.000523
12:19:04 T:123145308819456  NOTICE: ----- error: -2.001182, rr: 0.998524, prop: -0.001001, int: 0.000523
12:19:05 T:123145308819456  NOTICE: ----- error: -3.169526, rr: 0.997940, prop: -0.001585, int: 0.000523
12:19:06 T:123145308819456  NOTICE: ----- error: -3.133604, rr: 0.997958, prop: -0.001567, int: 0.000523
12:19:07 T:123145308819456  NOTICE: ----- error: -2.228050, rr: 0.998410, prop: -0.001114, int: 0.000523
12:19:08 T:123145308819456  NOTICE: ----- error: -1.540539, rr: 0.998754, prop: -0.000770, int: 0.000523
12:19:08 T:123145321213952   DEBUG: AddOnLog: VDR VNSI Client: Open - failed to connect to the backend (Operation timed out)
12:19:09 T:123145308819456  NOTICE: ----- error: -2.269183, rr: 0.998390, prop: -0.001135, int: 0.000523
12:19:10 T:123145308819456  NOTICE: ----- error: -3.035096, rr: 0.998007, prop: -0.001518, int: 0.000523
12:19:11 T:123145308819456  NOTICE: ----- error: -3.575782, rr: 0.997737, prop: -0.001788, int: 0.000523
12:19:12 T:123145308819456  NOTICE: ----- error: -2.274717, rr: 0.998387, prop: -0.001137, int: 0.000523
12:19:13 T:123145308819456  NOTICE: ----- error: -1.826530, rr: 0.998611, prop: -0.000913, int: 0.000523
12:19:14 T:123145308819456  NOTICE: ----- error: -2.622464, rr: 0.998213, prop: -0.001311, int: 0.000523
12:19:15 T:123145308819456  NOTICE: ----- error: -3.325181, rr: 0.997862, prop: -0.001663, int: 0.000523
12:19:16 T:123145308819456  NOTICE: ----- error: -4.446331, rr: 0.997301, prop: -0.002223, int: 0.000523
```

this is OSX where audio sink is not as accurate as ALSA. error stays in a band of +-1ms.

**Comment by MartijnKaijser** [comment]
ping @FernetMenta

**Comment by FernetMenta** [comment]
most of those changes does not make sense to me.

**Pull request closed without merging**